server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -144,10 +144,17 @@ public static State fromString(String state) {
     public static final Setting<Integer> INDEX_ROUTING_PARTITION_SIZE_SETTING =
             Setting.intSetting(SETTING_ROUTING_PARTITION_SIZE, 1, 1, Property.IndexScope);
 
+    private static String calculateNumOfRoutingShards(String parsedNumOfShards) {
+        final int maxNumShards = Integer.parseInt(System.getProperty("es.index.max_number_of_shards", "1024"));
+        int numOfShards = Integer.parseInt(parsedNumOfShards);
+        return String.valueOf(numOfShards * (maxNumShards/numOfShards));
+    }
+
     public static final Setting<Integer> INDEX_NUMBER_OF_ROUTING_SHARDS_SETTING =
         Setting.intSetting(
             "index.number_of_routing_shards",
             INDEX_NUMBER_OF_SHARDS_SETTING,
```

Comment:
Not sure, but should we replace current default ( INDEX_NUMBER_OF_SHARDS_SETTING) by new calculated value?

Reply:
@BaurzhanSakhariev In the next line `IndexMetadata::calculateNumOfRoutingShards` is passed as a function to to an overloaded `intSetting` method. Did you check it?

Reply:
Let's use [this ](https://github.com/crate/crate/blob/80bf834d0bc9d01d2b8daada92a606b1db808fe2/server/src/main/java/org/elasticsearch/common/settings/Setting.java#L1156-L1165) method for simplicity - then we wouldnt need another overload and also can just directly give calculated integer number as a parameter

Reply:
@BaurzhanSakhariev I saw that in the overriden validate method. Inside validate we can have access to settings. 
On the other hand when:
```
public static final Setting<Integer> INDEX_NUMBER_OF_ROUTING_SHARDS_SETTING_2 =
        Setting.intSetting(
            "index.number_of_routing_shards_2",
            ....
```
is called at compile time how can we access settings?

Could you please provide a more detailed example of how you would do this calculation?

Reply:
ah, sorry for the confusion. we cannot access `numOfShards` (I missed that we have settings in validator).

then in general your approach is fine (use some setting as a default) - the only my concern is usage of the trick:
take `numOfShards` as default but make parseInt a bit "smart" to actually take string and make some extra computations. While it might work, just a bit confusing for the reader to see 
```
Setting.intSetting(
            "index.number_of_routing_shards",
            INDEX_NUMBER_OF_SHARDS_SETTING, // might have an impression that this is the default value
            // but actual trick is here
          
```

Maybe just for the sake of readability (also to remove new override) we can use [this] (https://github.com/crate/crate/blob/7e140f7f7357e7d776d2c9cbb1a1b1e00153c73d/server/src/main/java/org/elasticsearch/common/settings/Setting.java#L216-L223)method. The main difference that it doesn't take some setting as a default but rather Function<Setting, String> which already hints that fallback setting will not be taken as is but rather be transformed.

Reply:
Np! So, I made some additional changes. 
Now when the user makes a query with `number_of_routing_shards` inside with:
`create table t1 (id int, thing text) with (number_of_routing_shards = 10)`

this value is stored and is used as a factor/multiplier when resharding.

When the user creates a table and does not provide `number_of_routing_shards` then this new calculation takes place.

So there are two new Settings and a check is made inside `MetadataCreateIndexService`:

```
if (recoverFromIndex == null) {
                    Settings idxSettings = indexSettingsBuilder.build();
                    idxSettings.get("number_of_routing_shards");
                    routingNumShards = idxSettings.get("index.number_of_routing_shards") != null ?
                        IndexMetadata.INDEX_NUMBER_OF_ROUTING_SHARDS_SETTING.get(idxSettings) :
                        IndexMetadata.INDEX_NUMBER_OF_ROUTING_SHARDS_DEFAULT_SETTING.get(idxSettings);
                } else {
```

The second "default" setting uses the first one as a fallback setting parameter:
```
public static final Setting<Integer> INDEX_NUMBER_OF_ROUTING_SHARDS_DEFAULT_SETTING =
        Setting.intSetting(
            "index.number_of_routing_shards",
            INDEX_NUMBER_OF_ROUTING_SHARDS_SETTING,
```

So the settings constructors I use seem suitable. 
Please do tell if you think they are not.

I am not very happy with the two similar settings but couldn't think of another way implementing this.

Reply:
Having second temporal setting looks hacky - could you please give a try to the approach I suggested in https://github.com/crate/crate/pull/12273#discussion_r837670508?

Reply:
I have discarded the second setting and used `calculateNumOfRoutingShards` method inside `MetadataCreateIndexService` similar to [this](https://github.com/elastic/elasticsearch/blob/898d84998b081d47ec6a9c2305f9bf4486d69be3/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java#L996)